crates/biome_js_semantic/src/events.rs

```diff
@@ -158,44 +152,84 @@ impl SemanticEvent {
 /// ```
 #[derive(Default, Debug)]
 pub struct SemanticEventExtractor {
+    // Event queue
     stash: VecDeque<SemanticEvent>,
+    /// Stack of scopes
     scopes: Vec<Scope>,
-    next_scope_id: usize,
-    /// At any point this is the set of available bindings and
-    /// the range of its declaration
+    // Number of generated scopes
+    // This allows assigning a unique scope id for every scope.
+    scope_count: usize,
+    /// At any point this is the set of available bindings in the current scope
     bindings: FxHashMap<BindingName, BindingInfo>,
+    // Type parameters bound in a `infer T` clause.
     infers: Vec<TsTypeParameterName>,
 }
 
-/// Holds the text range of the token when it is bound,
-/// along with the kind of declaration
-#[derive(Debug)]
-struct BindingInfo {
-    text_range: TextRange,
-    /// For export determination, it is necessary to provide
-    /// information on how a specific token is bound
-    declaration_kind: JsSyntaxKind,
+/// A declaration is either a type, a value, or both.
+///
+/// For example a class is botha  type and a value in TypeScript.
+/// A variable declaration is always a value, and a type alias is always a type.
+#[derive(Debug, Hash, Eq, PartialEq, Copy, Clone)]
+enum DeclarationKind {
+    Type,
+    Value,
+    Both,
 }
 
+/// A binding name is either a type or a value.
+///
+/// If a declaration is both a type and a value,
```

Comment:
The first sentence contrasts with the second because "either" means one OR the other, not both.